qiskit/circuit/quantumcircuit.py

```diff
@@ -3044,3 +2979,81 @@ def _compare_parameters(param1, param2):
 
     # else sort by name
     return _standard_compare(param1.name, param2.name)
+
+
+def _get_composite_circuit_qasm_from_instruction(
+    instruction, existing_gate_names, existing_composite_circuits
+):
+    """Returns OpenQASM string composite circuit given an instruction.
+    The given instruction should be the result of composite_circuit.to_instruction()."""
+
+    if instruction.definition is None:
+        raise ValueError(f'Instruction "{instruction.name}" is not defined.')
+
+    gate_parameters = ",".join(["param%i" % num for num in range(len(instruction.params))])
+    qubit_parameters = ",".join(["q%i" % num for num in range(instruction.num_qubits)])
+    composite_circuit_gates = ""
+
+    definition = instruction.definition
+    definition_bit_labels = {
+        bit: idx for bits in (definition.qubits, definition.clbits) for idx, bit in enumerate(bits)
+    }
+    for sub_instruction, qargs, _ in definition:
+        if sub_instruction.name not in existing_gate_names:
+            if sub_instruction not in existing_composite_circuits:
+                existing_composite_circuits.insert(0, sub_instruction)
+                _get_composite_circuit_qasm_from_instruction(
+                    sub_instruction, existing_gate_names, existing_composite_circuits
+                )
+        gate_qargs = ",".join(
+            ["q%i" % index for index in [definition_bit_labels[qubit] for qubit in qargs]]
+        )
+        composite_circuit_gates += "%s %s; " % (sub_instruction.qasm(), gate_qargs)
+
+    if composite_circuit_gates:
+        composite_circuit_gates = composite_circuit_gates.rstrip(" ")
+
+    if gate_parameters:
+        qasm_string = "gate %s(%s) %s { %s }" % (
+            instruction.name,
+            gate_parameters,
+            qubit_parameters,
+            composite_circuit_gates,
+        )
+    else:
+        qasm_string = "gate %s %s { %s }" % (
+            instruction.name,
+            qubit_parameters,
+            composite_circuit_gates,
+        )
+
+    return qasm_string
+
+
+def _insert_composite_gate_definition_qasm(
+    string_temp, existing_gate_names, existing_composite_circuits, extension_lib
+):
+    """Insert composite gate definition QASM code right after extension library in the header"""
+
+    gate_definition_string = ""
+
+    # Cycle through all gate definitions and add all undefined gates to the list
+    for instruction in existing_composite_circuits:
+        _get_composite_circuit_qasm_from_instruction(
+            instruction, existing_gate_names, existing_composite_circuits
+        )
```

Comment:
I'm a bit confused about this block, why is this needed? It seemed to me that the next block (L3047-3054) iterates over the same set of instructions 🤔

Reply:
There's probably better way to structure the code. Basically this loop is used to add all gates that are not in `existing_composite_circuits` to `existing_composite_circuits` list, while the loop below is to add qasm definition string to the output qasm string.

Reply:
Would it be simpler to do either this in one go or to have two separate functions for this?

Reply:
Yeah, now I think about it, I think it should be two separate functions. Let see if I can come up with a quick solution.

Reply:
I separated this function into two: `_add_sub_instruction_to_existing_composite_circuits` and `_get_composite_circuit_qasm_from_instruction` in the latest two commits. Can you please have a look? I hope the code now looks clearer and easier to understand.